core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -375,1083 +380,674 @@ public boolean isPreloaded() {
             });
    }
 
-   @Override
-   public void disableStore(String storeType) {
-      if (enabled) {
-         boolean noMoreStores;
-         storesMutex.writeLock().lock();
-         publisherSemaphore.acquireUninterruptibly(Integer.MAX_VALUE);
+   private Single<?> preloadKey(long flags, MarshallableEntry me, DataConversion keyDataConversion, DataConversion valueDataConversion) {
+      // CallInterceptor will preserve the timestamps if the metadata is an InternalMetadataImpl instance
+      InternalMetadataImpl metadata = new InternalMetadataImpl(me.getMetadata(), me.created(), me.lastUsed());
+      Object key = keyDataConversion.toStorage(me.getKey());
+      Object value = valueDataConversion.toStorage(me.getValue());
+      PutKeyValueCommand cmd = commandsFactory.wired().buildPutKeyValueCommand(key, value, keyPartitioner.getSegment(key), metadata, flags);
+      cmd.setInternalMetadata(me.getInternalMetadata());
+
+      CompletionStage<Object> stage;
+      if (configuration.transaction().transactionMode().isTransactional() && transactionManager != null) {
+         final Transaction transaction = suspendIfNeeded();
+         CompletionStage<Transaction> putStage;
          try {
-            removeCacheLoader(storeType, loaders);
-            removeCacheWriter(storeType, nonTxWriters);
-            removeCacheWriter(storeType, txWriters);
-            noMoreStores = loaders.isEmpty() && nonTxWriters.isEmpty() && txWriters.isEmpty();
-            readOnly = nonTxWriters.isEmpty() && txWriters.isEmpty();
-
-            if (!noMoreStores) {
-               // Immediately poll store availability as the disabled store may have been the cause of the unavailability
-               pollStoreAvailability();
-            }
-         } finally {
-            publisherSemaphore.release(Integer.MAX_VALUE);
-            storesMutex.writeLock().unlock();
+            beginIfNeeded();
```

Comment:
I don't understand why the `<suspend|begin|resume|commit>IfNeeded` methods are required?

`configuration.transaction().transactionMode().isTransactional() && transactionManager != null` is always  true when these methods are called. Can't we just execute the `transactionManager.` methods inline?

Reply:
Sadly this is how it was before. TBH I don't know and I didn't want to mess with this code :D

I can make a JIRA to reevaluate this though.

Reply:
It might be that it needed the blocking executor when transactions were blocking. We might be able to remove it now that I made that "non blocking" for the caller when in a tx.

Reply:
I created https://issues.redhat.com/browse/ISPN-11858